### PR TITLE
Adapt to upstream changes to return_types

### PR DESCRIPTION
**Pull request by aviatesk**
Fixes https://github.com/JuliaGPU/GPUCompiler.jl/issues/309

Still we want to wait for the release of [next beta version](https://github.com/JuliaLang/julia/pull/44324) and change the version compat accordingly.

### CODE REVIEW

**Comment by maleadt** [comment]
> Still we want to wait for the release of [next beta version](https://github.com/JuliaLang/julia/pull/44324) and change the version compat accordingly.

That will just be `v"1.8.0-beta2"`, right?

**Comment by aviatesk** [comment]
> That will just be `v"1.8.0-beta2"`, right?

yep, done!

File: src/validation.jl [hunk 1]
```diff
@@ -16,7 +16,12 @@ function check_method(@nospecialize(job::CompilerJob))
         cache = ci_cache(job)
         mt = method_table(job)
         interp = GPUInterpreter(cache, mt, job.source.world)
-        rt = Base.return_types(job.source.f, job.source.tt, interp)[1]
+        @static if VERSION ≥ v"1.8-beta2"
```
Comment (maleadt) [inline src/validation.jl:19]: Should we be more fine-grained here?

```suggestion
        @static if v"1.8-beta2" <= VERSION < v"1.9-" || VERSION ≥ v"1.9.0-DEV.190"
```

Comment (aviatesk) [reply src/validation.jl:19]: Thanks, I just ignored the possibility to use `v"1.9-" < v"1.9.0-DEV.190"`.

Comment (maleadt) [reply src/validation.jl]: Yeah that makes sense, but it does happen that we need to bisect Julia in combination with GPUCompiler, so it's better to have the version checks be as accurate as possible then.

**Comment by maleadt** [comment]
Tested this on 1.8-beta2 (current `release-1.8`), and now overrides are back to being broken 😢 

```
overrides: Test Failed at /home/tim/Julia/pkg/GPUCompiler/test/native.jl:387
  Expression: occursin("ret i64 1", ir)
   Evaluated: occursin("ret i64 1", ";  @ /home/tim/Julia/pkg/GPUCompiler/test/native.jl:380 within `kernel`\n; Function Attrs: alwaysinline\ndefine i64 @julia_kernel_4321() local_unnamed_addr #0 {\ntop:\n  ret i64 0\n}\n")
```

Specifically, the reproducer from https://github.com/JuliaLang/julia/issues/44174 fails again (it didn't on beta1, but constant propagation was pretty much disabled with overlayed method tables there).

**Comment by aviatesk** [comment]
Hmm... I will look into it.

**Comment by aviatesk** [comment]
> Specifically, the reproducer from [JuliaLang/julia#44174](https://github.com/JuliaLang/julia/issues/44174) fails again (it didn't on beta1, but constant propagation was pretty much disabled with overlayed method tables there).

I confirmed that should work now. Basically I forgot that I backported https://github.com/JuliaLang/julia/pull/44389 also.

**Comment by maleadt** [comment]
Works great, thanks!

**Pull request merged**